workflow/src/test/kotlin/com/r3/corda/sdk/token/workflow/statesAndContracts/DiamondGradingReport.kt

```diff
@@ -0,0 +1,70 @@
+package com.r3.corda.sdk.token.workflow.statesAndContracts
+
+import com.r3.corda.sdk.token.contracts.EvolvableTokenContract
+import com.r3.corda.sdk.token.contracts.states.EvolvableTokenType
+import net.corda.core.contracts.Contract
+import net.corda.core.contracts.UniqueIdentifier
+import net.corda.core.contracts.requireThat
+import net.corda.core.identity.Party
+import net.corda.core.serialization.CordaSerializable
+import net.corda.core.transactions.LedgerTransaction
+import java.math.BigDecimal
+
+/**
+ * The [DiamondGradingReport] is inspired by the grading reports issued by the Gemological Institute of America
+ * (GIA). For more excellent information on diamond grading, please see the (GIA's website)[http://www.gia.edu].
+ */
+class DiamondGradingReport : EvolvableTokenContract(), Contract {
+
+    override fun additionalCreateChecks(tx: LedgerTransaction) {
+        val outputDiamond = tx.outputsOfType<State>().first()
+        requireThat {
+            "Diamond's carat weight must be greater than 0 (zero)" using (outputDiamond.caratWeight > BigDecimal.ZERO)
+        }
+    }
+
+    override fun additionalUpdateChecks(tx: LedgerTransaction) {
+        val inDiamond = tx.outputsOfType<State>().first()
+        val outDiamond = tx.outputsOfType<State>().first()
+        requireThat {
+            "Diamond's carat weight may not be changed" using (inDiamond.caratWeight == outDiamond.caratWeight)
+            "Diamond's color may not be changed" using (inDiamond.color == outDiamond.color)
+            "Diamond's clarity may not be changed" using (inDiamond.clarity == outDiamond.clarity)
+            "Diamond's cut may not be changed" using (inDiamond.cut == outDiamond.cut)
+        }
+    }
+
+    @CordaSerializable
+    enum class ColorScale { A, B, C, D, E, F }
+
+    @CordaSerializable
+    enum class ClarityScale { A, B, C, D, E, F }
+
+    @CordaSerializable
+    enum class CutScale { A, B, C, D, E, F }
+
+    data class State(
+            val caratWeight: BigDecimal,
+            val color: ColorScale,
+            val clarity: ClarityScale,
+            val cut: CutScale,
+            val assessor: Party,
+            val requester: Party,
+            override val linearId: UniqueIdentifier = UniqueIdentifier()
+    ) : EvolvableTokenType() {
+        constructor(
+                caratWeight: String,
+                color: ColorScale,
+                clarity: ClarityScale,
+                cut: CutScale,
+                assessor: Party,
+                requester: Party,
+                linearId: UniqueIdentifier = UniqueIdentifier()) : this(BigDecimal(caratWeight), color, clarity, cut, assessor, requester, linearId)
+
+        override val maintainers get() = listOf(assessor)
+
+        override val participants get() = setOf(assessor, requester).toList()
```

Comment:
This is a slightly contrived example but given that it is possible to subclass EvolvableTokenType and change the definition of participants, I thought it may be interesting to explore what that would entail. This led to the logic in Create/Update for separating maintainers from participants.

In this example, the gem grading body has the unilateral right to issue and amend grading reports. For provenance, suppose the grading body also wanted to record publicly who requested the report; in other words, the grading body assessed the gem on behalf of the requester.

Since the requester was publicly recorded in the state, I added it to the participants list instead of setting up a separate subscription / distribution list.

Speaking of subscription / distribution list, given that there can be multiple maintainers, should we explore a more robust distribution model? I understand the current configuration records the subscription to one node only, which may mean some subscribers are not updated if maintainers change, a maintainer who does not hold the subscription makes an amendment, etc.

I was thinking a SubscriptionContract / Subscription that followed the EvolvableTokenType maintainers.

Reply:
Thanks - this all makes sense. I guess at the high level we have:

1. active participants (typically these parties will sign issues and updates)
2. named observers - also participants but cannot action anything 
3. un-named observers - these parties are not included in the state

We should probably support all three, even for tokens, as you have done here. 

Regarding distribution list - yes we really need to get data distribution groups out there! See the corda technical whitepaper for what they'll look like.